app/src/org/commcare/views/widgets/MediaWidget.java

```diff
@@ -0,0 +1,165 @@
+package org.commcare.views.widgets;
+
+import android.content.Context;
+import android.util.Log;
+import android.view.inputmethod.InputMethodManager;
+import android.widget.Button;
+import android.widget.LinearLayout;
+
+import org.commcare.activities.FormEntryActivity;
+import org.commcare.logic.PendingCalloutInterface;
+import org.javarosa.core.model.data.IAnswerData;
+import org.javarosa.core.model.data.IntegerData;
+import org.javarosa.core.model.data.InvalidData;
+import org.javarosa.core.model.data.StringData;
+import org.javarosa.form.api.FormEntryPrompt;
+
+import java.io.File;
+
+/**
+ * Generic logic for capturing or choosing audio/video/image media
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public abstract class MediaWidget extends QuestionWidget {
+    private static final String TAG = MediaWidget.class.getSimpleName();
+
+    protected Button mCaptureButton;
+    protected Button mPlayButton;
+    protected Button mChooseButton;
+    protected String mBinaryName;
+
+    protected final PendingCalloutInterface pendingCalloutInterface;
+    protected final String mInstanceFolder;
+
+    private int oversizedMediaSize;
+
+    public MediaWidget(Context context, FormEntryPrompt prompt,
+                       PendingCalloutInterface pendingCalloutInterface) {
+        super(context, prompt);
+
+        this.pendingCalloutInterface = pendingCalloutInterface;
+
+        mInstanceFolder =
+                FormEntryActivity.mInstancePath.substring(0,
+                        FormEntryActivity.mInstancePath.lastIndexOf("/") + 1);
+
+        setOrientation(LinearLayout.VERTICAL);
+        initializeButtons();
+        setupLayout();
+
+        loadAnswerFromDataModel();
+    }
+
+    private void loadAnswerFromDataModel() {
+        mBinaryName = mPrompt.getAnswerText();
+        if (mBinaryName != null) {
+            reloadFile();
+        } else {
+            checkForOversizedMedia(mPrompt.getAnswerValue());
+            togglePlayButton(false);
+        }
+    }
+
+    protected void reloadFile() {
+        togglePlayButton(true);
+        File f = new File(mInstanceFolder + mBinaryName);
+        checkFileSize(f);
+    }
+
+    protected void togglePlayButton(boolean enabled) {
+        mPlayButton.setEnabled(enabled);
+    }
+
+    protected abstract void initializeButtons();
+
+    protected void setupLayout() {
+        addView(mCaptureButton);
+        addView(mChooseButton);
+        addView(mPlayButton);
+    }
+
+    @Override
+    public IAnswerData getAnswer() {
+        if (mBinaryName != null) {
+            return new StringData(mBinaryName);
+        } else if (oversizedMediaSize > 0) {
```

Comment:
Shouldn't this check happen first? Or will `mBinaryName` always be null when `oversizedMediaSize > 0`?